src/js/touchevents.tsx

```diff
@@ -166,80 +174,91 @@ class TouchEventBoundary extends React.Component<TouchEventBoundaryProps> {
+          // the "check-label" if sentence, so we have to assign it to a variable here first
+          typeof this.props.labelName === 'string'
+          ? props[this.props.labelName]
           : undefined;
 
-      // For some reason type narrowing doesn't work as expected with indexing when checking it all in one go in
-      // the "check-label" if sentence, so we have to assign it to a variable here first
-      let labelValue;
-      if (typeof this.props.labelName === 'string')
-        labelValue = props?.[this.props.labelName];
-
-      // Check the label first
-      if (sentryLabel && !this._isNameIgnored(sentryLabel)) {
-        if (!activeLabel) {
-          activeLabel = sentryLabel;
-        }
-        componentTreeNames.push(sentryLabel);
-      } else if (
-        typeof labelValue === 'string' &&
-        !this._isNameIgnored(labelValue)
-      ) {
-        if (!activeLabel) {
-          activeLabel = labelValue;
-        }
-        componentTreeNames.push(labelValue);
-      } else if (currentInst.elementType) {
-        const { elementType } = currentInst;
-
-        if (
-          elementType.displayName &&
-          !this._isNameIgnored(elementType.displayName)
-        ) {
-          // Check display name
-          if (!activeDisplayName) {
-            activeDisplayName = elementType.displayName;
-          }
-          componentTreeNames.push(elementType.displayName);
-        }
+      if (typeof labelValue === 'string' && labelValue.length > 0) {
+        info.label = labelValue;
+      }
+
+      if (!info.name && currentInst.elementType?.displayName) {
+        info.name = currentInst.elementType?.displayName;
       }
 
+      this._pushIfNotIgnored(touchPath, info);
+
       currentInst = currentInst.return;
     }
 
-    const finalLabel = activeLabel ?? activeDisplayName;
-
-    if (componentTreeNames.length > 0 || finalLabel) {
-      this._logTouchEvent(componentTreeNames, finalLabel);
+    const label = touchPath.find(info => info.label)?.label;
+    if (touchPath.length > 0) {
+      this._logTouchEvent(touchPath, label);
     }
 
     this._tracingIntegration?.startUserInteractionTransaction({
-      elementId: activeLabel,
+      elementId: label,
       op: UI_ACTION_TOUCH,
     });
   }
+
+  /**
+   * Pushes the name to the componentTreeNames array if it is not ignored.
+   */
+  private _pushIfNotIgnored(touchPath: TouchedComponentInfo[], value: TouchedComponentInfo): boolean {
+    if (!value.name && !value.label) {
+      return false;
+    }
+    if (value.name && this._isNameIgnored(value.name)) {
+      return false;
+    }
+    if (value.label && this._isNameIgnored(value.label)) {
+      return false;
+    }
+
+    // Deduplicate same subsequent items.
+    if (touchPath.length > 0 && JSON.stringify(touchPath[touchPath.length - 1]) === JSON.stringify(value)) {
+      return false;
+    }
```

Comment:
Was this happening a lot in your testing @vaind ?

Let's add a test for it.

Reply:
it did. And it was tested as part of another test case but I've added a custom one for it.

Reply:
Aaa, thanks, I've missed that. Thanks for splitting it.